go/types/subtype.go

```diff
@@ -105,73 +124,111 @@ func isSubtype(requiredType, concreteType *Type, allowExtraStructFields bool, pa
+func compoundSubtype(requiredType, concreteType *Type, hasExtra bool, parentStructTypes []*Type) (bool, bool) {
 	// If the concrete type is a union then all the types in the union must be subtypes of the required typ. This also means that a compound type with an empty union is going to be a subtype of all compounds, List<> is a subtype of List<T> for all T.
 	if concreteType.TargetKind() == UnionKind {
 		for _, ct := range concreteType.Desc.(CompoundDesc).ElemTypes {
-			if !isSubtype(requiredType, ct, allowExtraStructFields, parentStructTypes) {
-				return false
+			isSub, hasExtra1 := isSubtypeDetails(requiredType, ct, hasExtra, parentStructTypes)
+			if !isSub {
+				return false, hasExtra1
 			}
 		}
-		return true
+		return true, hasExtra
 	}
-	return isSubtype(requiredType, concreteType, allowExtraStructFields, parentStructTypes)
+	return isSubtypeDetails(requiredType, concreteType, hasExtra, parentStructTypes)
 }
 
-// IsValueSubtypeOf returns whether a value is a subtype of a type.
 func IsValueSubtypeOf(v Value, t *Type) bool {
+	isSub, _ := isValueSubtypeOfDetails(v, t, false)
+	return isSub
+}
+
+// IsValueSubtypeOfDetails returns two values:
+//   isSub - which indicates whether v is a subtype of t.
+//   hasExtra - which indicates whether v has additional fields. This field has
+//              no meaning if IsSub is false.
+//
+// For example, given the following data:
+//   type1 := struct S {               v := Struct S1 {
+//       a Number | string                 a: "hello"
+//       b ?int                            b: 2
+//   }                                 }
+// IsValueSubtypeOfDetails(v, type1) would return isSub == true, and hasExtra == false
+//
+// And given these types:
+//   type2 := struct S {               v := Struct S1 {
+//       a Number | string                 a: "hello"
+//       b ?int                            b: 2
+//   }                                     c: "hello again"
+//                                     }
+// IsValueSubtypeOfDetails(v, type1) would return isSub == true, and hasExtra == true
+func IsValueSubtypeOfDetails(v Value, t *Type) (bool, bool) {
+	return isValueSubtypeOfDetails(v, t, false)
+}
+
+func isValueSubtypeOfDetails(v Value, t *Type, hasExtra bool) (bool, bool) {
 	switch t.TargetKind() {
 	case BoolKind, NumberKind, StringKind, BlobKind, TypeKind:
-		return v.Kind() == t.TargetKind()
+		return v.Kind() == t.TargetKind(), hasExtra
 	case ValueKind:
-		return true
+		return true, hasExtra
 	case UnionKind:
+		var anonStruct *Type
+
 		for _, et := range t.Desc.(CompoundDesc).ElemTypes {
-			if IsValueSubtypeOf(v, et) {
-				return true
+			if et.TargetKind() == StructKind && et.Desc.(StructDesc).Name == "" {
+				anonStruct = et
+				continue
+			}
+			isSub, hasMore := isValueSubtypeOfDetails(v, et, hasExtra)
+			if isSub {
+				hasExtra = hasExtra || hasMore
+				return isSub, hasExtra
 			}
 		}
-		return false
+
+		if anonStruct != nil {
```

Comment:
Copied a comment from elsewhere in the source that explains what is going on here. Also, we verified that there can only be one anon struct per type.